File: src/main/java/com/will/utils/xml/XMLObject.java
package com.will.utils.xml;

import java.io.Serializable;
import java.util.*;

/**
 * Created by Will on 2016/10/18 9:49.
 */
public class XMLObject implements Map<String,Object>,Cloneable, Serializable {
    private static final long serialVersionUID = -7006835342455866809L;
    private static final int DEFAULT_INITIAL_CAPACITY = 16;
    private final Map<String, Object> map;

    public XMLObject() {
        this(DEFAULT_INITIAL_CAPACITY, false);
    }

    public XMLObject(Map<String, Object> map) {
        this.map = map;
    }

    public XMLObject(boolean ordered) {
        this(DEFAULT_INITIAL_CAPACITY, ordered);
    }

    public XMLObject(int initialCapacity) {
        this(initialCapacity, false);
    }

    public XMLObject(int initialCapacity, boolean ordered) {
        if(ordered) {
            this.map = new LinkedHashMap(initialCapacity);
        } else {
            this.map = new HashMap(initialCapacity);
        }

    }
    @Override
    public int size() {
        return this.map.size();
    }

    @Override
    public boolean isEmpty() {
        return this.map.isEmpty();
    }

    @Override
    public boolean containsKey(Object key) {
        return this.map.containsKey(key);
    }

    @Override
    public boolean containsValue(Object value) {
        return this.map.containsKey(value);
    }

    @Override
    public Object get(Object key) {
        return this.map.get(key);
    }

    public XMLObject getXMLObject(String key) {
        Object value = this.map.get(key);
        if(value instanceof XMLObject){
            return (XMLObject) value;
        }else{
            throw new RuntimeException("List 不能转换为XMLObject");
        }
    }

    public List<XMLObject> getList(String key){
        Object value = this.map.get(key);
        if(value instanceof List){
            return (List) value;
        }else{
            throw new RuntimeException("XMLObject 不能转换为List");
        }
    }

    @Override
    public Object put(String key, Object value) {
        return this.map.put(key,value);
    }

    @Override
    public Object remove(Object key) {
        return this.map.remove(key);
    }

    @Override
    public void putAll(Map<? extends String, ?> m) {
        this.map.putAll(m);
    }

    @Override
    public void clear() {
        this.map.clear();
    }

    @Override
    public Set<String> keySet() {
        return this.map.keySet();
    }

    @Override
    public Collection<Object> values() {
        return this.map.values();
    }

    @Override
    public Set<Entry<String, Object>> entrySet() {
        return this.map.entrySet();
    }

    public Object clone() {
        return new XMLObject(new HashMap(this.map));
    }

    public boolean equals(Object obj) {
        return this.map.equals(obj);
    }

    public int hashCode() {
        return this.map.hashCode();
    }
}

File: src/main/java/com/will/utils/xml/XMLObject.kt
package com.will.utils.xml

import java.io.Serializable

/**
 * Created by zoumy on 2017/5/11 13:58.
 */
class XMLObject : MutableMap<String,Any>,Serializable {

    val VALUE:String = "value"
    val DEFAULT_INITIAL_CAPACITY:Int = 16
    val map:MutableMap<String,Any>

    constructor(initialCapacity:Int,ordered:Boolean = false){
        if (ordered){
            this.map = LinkedHashMap(initialCapacity)
        }else{
            this.map = HashMap(initialCapacity)
        }
    }

    constructor():this(16)

    constructor(ordered: Boolean):this(16,ordered)

    constructor(map:Map<String,Any>){
        this.map = map as MutableMap<String, Any>
    }

    constructor(initialCapacity: Int):this(initialCapacity,false)

    override val size: Int
        get() = this.map.size

    override fun isEmpty(): Boolean {
        return this.map.isEmpty()
    }

    override fun containsKey(key: String): Boolean {
        return this.map.containsKey(key)
    }

    override fun containsValue(value: Any): Boolean {
        return this.map.containsValue(value)
    }

    override fun get(key: String): Any? {
        return this.map[key]
    }

    override val entries: MutableSet<MutableMap.MutableEntry<String, Any>>
        get() = this.map.entries
    override val keys: MutableSet<String>
        get() = this.map.keys
    override val values: MutableCollection<Any>
        get() = this.map.values

    override fun clear() {
        this.map.clear()
    }

    override fun put(key: String, value: Any): Any? {
        return this.map.put(key,value)
    }

    override fun putAll(from: Map<out String, Any>) {
        this.map.putAll(from)
    }

    override fun remove(key: String): Any? {
        return this.map.remove(key)
    }



    /**
     * 根据Key获取一个XMLObject 可能会空，可能会报类型转换异常

     * @param key the key
     * *
     * @return the xml object
     */
    fun getXMLObject(key: String): XMLObject {
        val value = this.map[key]
        if (value is XMLObject || value == null) {
            return value as XMLObject
        } else {
            throw ClassCastException(value.javaClass.name + "不能转换为XMLObject")
        }
    }


    /**
     * 根据Key获取一个List&lt;XMLObject&gt; 可能会空，可能会报类型转换异常

     * @param key the key
     * *
     * @return the list
     * *
     * @Auth Will
     * *
     * @Date 2017 -03-27 10:09:19
     */
    fun getList(key: String): List<XMLObject> {
        val value = this.map[key]
        if (value is List<*> || value == null) {
            return value as List<XMLObject>
        } else {
            throw ClassCastException(value.javaClass.name + " 不能转换为List")
        }
    }


    /**
     * 获取一个XMLObject对象下的value

     * @return the value
     */
    fun getValue(): Any? {
        return this.map[VALUE]
    }


    /**
     * 设置XMLObject对象下的value值

     * @param value the value
     * *
     * @return the object
     * *
     * @Auth Will
     * *
     * @Date 2017 -03-28 09:39:21
     */
    fun setValue(value: Any): Any? {
        return this.map.put(VALUE, value)
    }

    /**
     * 递归获取某节点下所有子集

     * @param key 关键字
     * *
     * @return 返回的结果
     */
    fun recursionValue(key: String): Any? {
        var o: Any? = null
        for ((key1, value) in this) {
            if (key1 == key) {
                o = value
                break
            } else {
                if (value is XMLObject) {
                    o = value.recursionValue(key)
                }
            }
        }
        return o
    }

    override fun toString(): String {
        var sb:StringBuilder = StringBuilder()
        sb.append("{")
        this.map.forEach { t, u ->
            sb.append("$t:$u\n")
        }
        sb.append("}")
        return sb.toString()
    }

}